Add scalar multiplication for `DirecVector`

skaunov opened the pull request:
Type value limit is left untested, so the `impl` should default to behavior across the crate on overflow.

Also I didn't learned macros yet, so I hope to return to this one to implement `Mul` for `Integer` types for commutativity.

here's snippet I had to withhold at this point
```
// impl<T> std::ops::Mul<&DirecVector<T>> for Integer
//     where T: std::ops::Mul<Output=T> + Copy
// {
//     type Output = DirecVector<T> ;

//     fn mul(self, rhs: T) -> Self::Output {
//         DirecVector{magnitude: rhs.magnitude * self, ..*rhs}
//     }
// }

// TODO return here when I learn macro https://stackoverflow.com/a/70220689
    // #[test]
    // fn zero_mul_direcvec() {
    //     let t = 0 * &DirecVector{ magnitude: -3, direction: Direction::Up };
    //     assert_eq!(t, DirecVector{ magnitude: 0, direction: Direction::Down });
    // }
```

skaunov committed:
Add scalar multiplication for `DirecVector`

brunoczim commented on src/direc.rs:
```diff
@@ -164,6 +180,18 @@ pub struct DirecVector<T> {
+/// A scalar multiplication for a vector.
+impl<T> std::ops::Mul<T> for &DirecVector<T>
+where
+    T: std::ops::Mul<Output = T> + Copy,
```
I think `Copy` is too restrictive. You could replace it with a `T: Clone` constraint instead.

skaunov replied:
I absolutely agree; even already made the change locally (AFAIU for copy-types `clone` just calls the same copy method). Apparently this modification would be better from theoretically (so why not do it better); and still I try to justify it with some example...

After some thinking let me put it following way. _Why `DirecVector` `magnitude` is *so generic*?_ Is it just "because you can,"  X) or any reasons behind it? Because I feel like I would make it `Integer` and I really would like to nurture some better intuition in me than I have now. X)

After some additional thinking I understand that it's exactly rationale why I used `Copy` back then. "All realistic scalar types are boil down to a copy-type in Rust" and I still can't come up with a sane example of scalar user would want to use that isn't copy. :exploding_head:

brunoczim commented on src/direc.rs:
```diff
@@ -164,6 +180,18 @@ pub struct DirecVector<T> {
 }
 
+/// A scalar multiplication for a vector.
+impl<T> std::ops::Mul<T> for &DirecVector<T>
```
Besides implementing `Mul<T> for &DirecVector<T>` , one could also implements `Mul<&T> for &DirecVector<T>`, also `Mul<T> for &DirecVector<T>`  and `Mul<T> for DirecVector<T>`  (in the latter case you don't even need a `Clone`). But this is optional, don't worry, I would implement it myself in case you don't.

skaunov replied:
Am I correct that all of them are indeed wrappers over `Mul<&T> for &DirecVector<T>`?

brunoczim reviewed (changes requested):
Well, I think that there is a more generalizing way of implementing it. Look at how it was implemented for `Vec2` ([implementation here](https://github.com/brunoczim/gardiz/blob/ca551aecf945b9fc008184a7ff25628aeb37fc83/src/coord.rs#L777-L942)). HOWEVER, this implementation involves macros, HOWEVER, I believe those macros are simple and are a good introduction for a beginner.

I would also add implementations for scalar addition, subtraction and division, using macros. Despite that, this is completely optional, you can do it in this PR, you can do it in a later PR, or you can just do what you are comfortable doing, as I can complete missing impls or refactor them with macros. Just let me know what you intend to do :D

skaunov commented:
Thanks a lot for comments --- will go through them! I have the VM image 
with the fork in semi-archived state. X) So there will be a time gap 
while I will be flying over everything there.

On Mon, Jul 3 2023 at 07:56:43 AM -0700, brunoczim 
***@***.***> wrote:
> ***@***.**** requested changes on this pull request.
> 
> Well, I think that there is a more generalizing way of implementing 
> it. Look at how it was implemented for Vec2 (implementation here 
> <https://github.com/brunoczim/gardiz/blob/ca551aecf945b9fc008184a7ff25628aeb37fc83/src/coord.rs#L777-L942>). 
> HOWEVER, this implementation involves macros, HOWEVER, I believe 
> those macros are simple and are a good introduction for a beginner.
> 
> I would also add implementations for scalar addition, subtraction and 
> division, using macros. Despite that, this is completely optional, 
> you can do it in this PR, you can do it in a later PR, or you can 
> just do what you are comfortable doing, as I can complete missing 
> impls or refactor them with macros. Just let me know what you intend 
> to do :D
> 
> In src/direc.rs 
> <https://github.com/brunoczim/gardiz/pull/2#discussion_r1250959614>:
> 
> > @@ -164,6 +180,18 @@ pub struct DirecVector<T> {
>      pub direction: Direction,
>  }
> 
> +/// A scalar multiplication for a vector.
> +impl<T> std::ops::Mul<T> for &DirecVector<T>
> +where
> +    T: std::ops::Mul<Output = T> + Copy,
> I think Copy is too restrictive. You could replace it with a T: Clone 
> constraint instead.
> 
> In src/direc.rs 
> <https://github.com/brunoczim/gardiz/pull/2#discussion_r1250982794>:
> 
> > @@ -164,6 +180,18 @@ pub struct DirecVector<T> {
>      pub direction: Direction,
>  }
> 
> +/// A scalar multiplication for a vector.
> +impl<T> std::ops::Mul<T> for &DirecVector<T>
> Besides implementing Mul<T> for &DirecVector<T> , one could also 
> implements Mul<&T> for &DirecVector<T>, also Mul<T> for 
> &DirecVector<T> and Mul<T> for DirecVector<T> (in the latter case you 
> don't even need a Clone). But this is optional, don't worry, I would 
> implement it myself in case you don't.
> 
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/brunoczim/gardiz/pull/2#pullrequestreview-1511296598>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/APXLOT3HCDIU5A4BHKN66SLXOLMSXANCNFSM6AAAAAAWNN4LUI>.
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>
>

skaunov commented:
Btw, have you ever got an idea to divide <src/coord.rs> further into sub`mod`s? Or is it a bad approach?

skaunov committed:
Add LHS and address two comments

Left hand side `Mul` is done here as [intended initially]; should go
over [the hint] to find possible improvements.
[intended initially]: https://github.com/brunoczim/gardiz/pull/2#issue-1647816420
[the hint]: https://github.com/brunoczim/gardiz/pull/2#pullrequestreview-1511296598

[Ease of restriction] to `Clone` is addressed.
[Ease of restriction]: https://github.com/brunoczim/gardiz/pull/2#discussion_r1250959614

[`impl` for refs] is addressed. Note that there's question pending,
which could be followed by adding full scope of relevant methods. Also
LHS macro should receive respective `impl`s.
[`impl` for refs]: https://github.com/brunoczim/gardiz/pull/2#discussion_r1250982794

brunoczim commented:
> Btw, have you ever got an idea to divide <src/coord.rs> further into submods? Or is it a bad approach?

I am not sure if anyone would agree with me, but I dislike the idea of splitting `Vec2` impls into multiple `mod`. But I think it is OK to split `coord.rs` if we add multiple items.
